### DISCUSSION TITLE
RLS policies that use headers dont seem to work on storage.objects

### DISCUSSION THREAD

**Discussion by surajshah** [supabase/supabase#14586] [Questions]
We are trying to use the"anon" access from a js application to store files in the storage:

 const options = {
  db: {
    schema: 'public',
  },
  auth: {
    autoRefreshToken: true,
    persistSession: true,
    detectSessionInUrl: true
  },
  global: {
    headers: { 'devicekey': 'ad62fde3-1ad8-5612-8f35-69f2cc600bbf-f01898ef5721' },
  },
}
const _supabase = supabase.createClient(SUPABASE_URL, SUPABASE_KEY, options);

var img = dataURLtoFile('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIAQMAAAD+wSzIAAAABlBMVEX///+/v7+jQ3Y5AAAADklEQVQI12P4AIX8EAgALgAD/aNpbtEAAAAASUVORK5CYII','test.png');
	const { data, error } = await _supabase.storage
            			.from('mybucket')
            			.upload('events/123456.png', img);

We now wanted to setup a policy where storing is allowed only if a client has a certain custom header (devicekey). 

And it seems that this doesn't work with storage.objects table even for certain standard headers like 'user-agent'.

Here is the policy we tried with:

**This works:**
CREATE POLICY "storage_objects_policy" ON storage.objects FOR ALL USING ((bucket_id = 'mybucket') and true);

**This doesn't work:**
CREATE POLICY "storage_objects_policy" ON storage.objects FOR ALL 
USING ((bucket_id = 'mybucket') and current_setting('request.headers', true):: json->>'user-agent'='Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36');

It gives this error in the client side:
http 401: new row violates row-level security policy for table \"objects\"

Note: for saving data to "normal" supabase tables, which we created in supabase, the same kind of "header" based policy works fine!

We wanted to know if anybody knows if using "headers" is supported for RLS policies in "storage.objects" or not?

Thanks!

**Comment by GaryAustin1** [comment]
https://github.com/supabase/storage-api/issues/216
Towards the end of that it says it did not work for hosted Supabase, and then Apr 4 updates says works.
I verified it does not work and did a pause/restore in case that loads the latest image of the storage-api.

**Comment by surajshah** [comment]
Hi,Thanks for the update, i will have it tested and update you.I would like to know, as this feature to pass the headers, seems "young" (so it may not be a mainstream feature) is here to stay or if it may get changed in future?I ask this because our solution will "heavily" use these headers, so we are concerned if it may lead to problems in the future?Our requirement is to make anonymous calls to supabase but passing a unique preshared "passkey" for each client that will be stored in a local setting in the client.Could you suggest us on this?Thanks in advance,SurajSent from my iPhoneOn 25 May 2023, at 17:50, GaryAustin1 ***@***.***> wrote:﻿
supabase/storage-api#216
Towards the end of that it says it did not work for hosted Supabase, and then Apr 4 updates says works.
I verified it does not work and did a pause/restore in case that loads the latest image of the storage-api.

**Reply by GaryAustin1** [reply]
You should ask in that issue I linked. I'm not a Supabase employee.

**Reply by surajshah** [reply]
Hi Gary, sorry about the misunderstanding. I will ask that in the link. Thank you for the suggestion!

**Comment by surajshah** [comment]
I tried to restart the project and the database via the button on this page:
https://supabase.com/dashboard/project/ofgpclxbmnewiswlvtmy/settings/general

Even after the restart it doesnt work! :(

**Reply by GaryAustin1** [reply]
Yeah, I tried the same thing with the same result.

**Reply by surajshah** [reply]
HI Gary,
I posted the question on the [supabase/storage-api#216](https://github.com/supabase/storage-api/issues/216) link as well.
Lets hope we can get some suggestion, because this is really essential for us to enforce our security policies.
Thanks
Suraj

**Comment by surajshah** [comment]
Hi All, I wrote to the supabase team and found that the project needs to run on "Storage V3", once they did it it automagically started working! Thank you supabase team!
